sources/api/migration/migrations/v1.4.0/registry-mirror-representation/src/main.rs

```diff
@@ -0,0 +1,126 @@
+            .map(|(k, v)| (k.to_owned(), v.to_owned()))
+            .collect();
+        let mut new_mirrors = Vec::new();
+        for (setting, endpoint) in mirrors {
+            // Need to split the full setting name to get the registry name.
+            // We get the index of the third '.' and split the string around there and leave out the '.'.
+            // Trim quotation marks around the registry name.
+            let registry = setting
+                .match_indices('.')
+                .nth(2)
+                .map(|(idx, _)| setting.split_at(idx + 1))
+                .unwrap_or_default()
+                .1
+                .trim_matches('"');
+            let mut registry_mirrors = Map::new();
+            registry_mirrors.insert("registry".to_string(), Value::String(registry.to_string()));
+            registry_mirrors.insert("endpoint".to_string(), endpoint.to_owned());
+            new_mirrors.push(Value::Object(registry_mirrors));
+            if let Some(data) = input.data.remove(&setting) {
+                println!("Removed setting '{}', which was set to '{}'", setting, data);
+            }
+        }
+        let data = Value::Array(new_mirrors);
+        println!(
+            "Creating new setting '{}', which is set to '{}'",
+            MIRRORS_SETTING_NAME, &data
+        );
+        input.data.insert(MIRRORS_SETTING_NAME.to_string(), data);
+        Ok(input)
+    }
+
+    /// Older versions store `settings.container-registry.mirrors` as `HashMap<SingleLineString, Vec<Url>>`.
+    /// Need to convert from `Vec<RegistryMirrors>`.
+    fn backward(&mut self, mut input: MigrationData) -> Result<MigrationData> {
+        if let Some(data) = input.data.get_mut(MIRRORS_SETTING_NAME).cloned() {
+            match data {
+                Value::Array(arr) => {
+                    if let Some(data) = input.data.remove(MIRRORS_SETTING_NAME) {
+                        println!(
+                            "Removed setting '{}', which was set to '{}'",
+                            MIRRORS_SETTING_NAME, data
+                        );
+                    }
+                    'reg_mirror: for obj in arr {
+                        if let Some(obj) = obj.as_object() {
+                            if let (Some(registry), Some(endpoint)) = (
+                                obj.get("registry").and_then(|s| s.as_str()),
+                                obj.get("endpoint"),
+                            ) {
+                                // Ensure the registry contains valid datastore key characters.
+                                // If we encounter any invalid key characters, we skip writing out
+                                // the setting key to prevent breakage of the datastore.
+                                for c in registry.chars() {
+                                    if !Key::valid_character(c) && c != KEY_SEPARATOR {
```

Comment:
Rather than the `for` and needing a label, I think you can use something like `registry.chars().all(|c| valid_character(c) || c == KEY_SEPARATOR)`?